File: packages/js/src/adf/validate.ts
import type { AdfDocument } from "./types.js"
import { isAdfDocument } from "./coerce.js"
import type { AnySchema } from "ajv-draft-04/dist/index.js"
import { existsSync, readFileSync } from "node:fs"
import { createRequire } from "node:module"
import { dirname, resolve } from "node:path"
import { fileURLToPath } from "node:url"

type ParseAdfOptions = {
  validateAdf?: boolean
}

type SchemaError = {
  instancePath?: string
  keyword?: string
  params?: Record<string, unknown>
  schema?: unknown
  schemaPath?: string
  message?: string
}

export type ValidationResult = {
  valid: boolean
  errors: string[]
}

const schema = loadPinnedAdfSchema()
const require = createRequire(import.meta.url)
const Ajv = require("ajv-draft-04")
  .default as typeof import("ajv-draft-04/dist/index.js").default
const ajv = new Ajv({ allErrors: true, strict: false })
const validatePinnedAdfSchema = ajv.compile(schema as AnySchema)

/**
 * Parses an unknown value as an ADF document, optionally validating it against
 * the pinned schema before returning the typed document.
 */
export function parseAdf(
  value: unknown,
  options: ParseAdfOptions = {},
): AdfDocument {
  if (isAdfDocument(value)) {
    if (options.validateAdf !== false) {
      const valid = validatePinnedAdfSchema(value)
      if (!valid) {
        throw new Error(
          formatSchemaError(validatePinnedAdfSchema.errors ?? [], value),
        )
      }
    }
    return value as AdfDocument
  }

  throw new Error("Invalid ADF root: expected doc version 1.")
}

/**
 * Validates an unknown value as ADF and returns a structured success/error
 * result instead of throwing.
 */
export function validateAdf(
  value: unknown,
  options: ParseAdfOptions = {},
): ValidationResult {
  try {
    parseAdf(value, options)
    return { valid: true, errors: [] }
  } catch (error) {
    return {
      valid: false,
      errors: [error instanceof Error ? error.message : "Invalid ADF."],
    }
  }
}

/**
 * Finds and loads the pinned ADF schema from either source or built package
 * layouts.
 */
function loadPinnedAdfSchema(): unknown {
  const here = dirname(fileURLToPath(import.meta.url))
  const candidates = [
    resolve(here, "../../schemas/adf-schema.json"),
    resolve(here, "../../../../schemas/adf-schema.json"),
    resolve(here, "../../../../../schemas/adf-schema.json"),
  ]

  for (const candidate of candidates) {
    if (existsSync(candidate)) {
      return JSON.parse(readFileSync(candidate, "utf8")) as unknown
    }
  }

  throw new Error("Pinned ADF JSON Schema not found.")
}

/**
 * Converts AJV schema errors into a short human-readable validation message.
 */
function formatSchemaError(errors: SchemaError[], value: unknown): string {
  const error = mostSpecificError(errors, value)
  const path =
    error.instancePath && error.instancePath.length > 0
      ? error.instancePath
      : ""
  const location = path.length > 0 ? ` at ${path}` : ""

  switch (error.keyword) {
    case "required": {
      const missing = String(error.params?.missingProperty ?? "unknown")
      return `Invalid ADF${location}: missing required property '${missing}'.`
    }
    case "additionalProperties": {
      const property = String(error.params?.additionalProperty ?? "unknown")
      return `Invalid ADF${location}: unexpected property '${property}'.`
    }
    case "enum":
      return `Invalid ADF${location}: value is not allowed.`
    case "type":
      return `Invalid ADF${location}: expected ${String(error.params?.type ?? "valid value")}.`
    case "minLength":
      return `Invalid ADF${location}: expected a non-empty string.`
    case "minItems":
      return `Invalid ADF${location}: expected at least one item.`
    default:
      return `Invalid ADF${location}: failed schema constraint '${error.keyword ?? "unknown"}'.`
  }
}

/**
 * Picks the schema error that most closely describes the invalid ADF location.
 */
function mostSpecificError(errors: SchemaError[], value: unknown): SchemaError {
  const relevantErrors = errors.filter(
    (error) => error.keyword !== "anyOf" && error.keyword !== "oneOf",
  )
  const candidates = relevantErrors.length > 0 ? relevantErrors : errors
  const matchingTypeError = candidates.find((error) =>
    errorMatchesActualType(error, value),
  )
  if (matchingTypeError) return matchingTypeError

  return candidates.reduce<SchemaError>(
    (best, error) =>
      errorSpecificity(error) > errorSpecificity(best) ? error : best,
    candidates[0] ?? {},
  )
}

/**
 * Checks whether a required-property error applies to the node or mark type at
 * the reported path.
 */
function errorMatchesActualType(error: SchemaError, value: unknown): boolean {
  if (error.keyword !== "required" || !error.schemaPath) return false

  const definition = /^#\/definitions\/([^/]+)\//.exec(error.schemaPath)?.[1]
  if (!definition) return false

  const expectedType = definition.replace(/_(mark|node)$/, "")
  const parentPath = (error.instancePath ?? "").replace(/\/attrs$/, "")
  const parent = valueAtPointer(value, parentPath)
  return (
    typeof parent === "object" &&
    parent !== null &&
    (parent as { type?: unknown }).type === expectedType
  )
}

/**
 * Resolves a JSON Pointer against a value so schema errors can be matched to
 * the actual ADF node.
 */
function valueAtPointer(value: unknown, pointer: string): unknown {
  if (pointer === "") return value
  return pointer
    .slice(1)
    .split("/")
    .map((part) => part.replace(/~1/g, "/").replace(/~0/g, "~"))
    .reduce<unknown>((current, part) => {
      if (Array.isArray(current)) return current[Number(part)]
      if (typeof current === "object" && current !== null) {
        return (current as Record<string, unknown>)[part]
      }
      return undefined
    }, value)
}

/**
 * Scores schema errors by path depth so deeper, more specific failures win.
 */
function errorSpecificity(error: SchemaError): number {
  const base = pathDepth(error.instancePath ?? "")
  return error.keyword === "required" &&
    (error.instancePath ?? "").includes("/attrs")
    ? base + 1
    : base
}

/**
 * Counts the number of JSON Pointer path segments.
 */
function pathDepth(path: string): number {
  return path === "" ? 0 : path.split("/").length
}

File: packages/python/src/md_adf/adf/validate.py
from __future__ import annotations

from dataclasses import dataclass, field
import json
from pathlib import Path
from typing import Any, cast

from jsonschema import Draft4Validator
from jsonschema.exceptions import ValidationError

from .coerce import is_adf_document
from .types import AdfDocument


@dataclass(frozen=True)
class ParseAdfOptions:
    """Options controlling how unknown values are parsed as ADF."""

    validate_adf: bool = True


@dataclass(frozen=True)
class ValidationResult:
    """Structured success/error result for ADF validation."""

    valid: bool
    errors: list[str] = field(default_factory=list)


def parse_adf(value: Any, options: Any = None) -> AdfDocument:
    """Parse an unknown value as an ADF document, optionally validating schema."""

    if is_adf_document(value):
        if _validate_enabled(options):
            errors = list(_VALIDATOR.iter_errors(value))
            if errors:
                raise ValueError(_format_schema_error(_most_specific_error(errors, value)))
        return cast(AdfDocument, value)

    raise ValueError("Invalid ADF root: expected doc version 1.")


def validate_adf(value: Any, options: Any = None) -> ValidationResult:
    """Validate an unknown value as ADF without raising validation errors."""

    try:
        parse_adf(value, options)
    except ValueError as exc:
        return ValidationResult(valid=False, errors=[str(exc)])

    return ValidationResult(valid=True)


def _schema_path() -> Path:
    """Find the pinned ADF JSON Schema in source or packaged layouts."""

    here = Path(__file__).resolve()
    candidates = [
        here.parents[1] / "schemas" / "adf-schema.json",
        here.parents[5] / "schemas" / "adf-schema.json",
    ]
    for candidate in candidates:
        if candidate.exists():
            return candidate

    raise RuntimeError("Pinned ADF JSON Schema not found.")


_SCHEMA = json.loads(_schema_path().read_text(encoding="utf-8"))
_VALIDATOR = Draft4Validator(_SCHEMA)


def _validate_enabled(options: Any) -> bool:
    """Read the validate_adf option from dataclass or mapping-style options."""

    if options is None:
        return True
    if isinstance(options, ParseAdfOptions):
        return options.validate_adf
    if hasattr(options, "validate_adf"):
        return bool(options.validate_adf)
    value = options.get("validate_adf", options.get("validateAdf", True))
    return bool(value)


def _format_schema_error(error: ValidationError) -> str:
    """Convert a jsonschema validation error into a concise user message."""

    path = _json_pointer(error.absolute_path)
    location = f" at {path}" if path else ""

    if error.validator == "required":
        missing = _required_property(error.message)
        return f"Invalid ADF{location}: missing required property '{missing}'."
    if error.validator == "additionalProperties":
        property_name = _additional_property(error.message)
        return f"Invalid ADF{location}: unexpected property '{property_name}'."
    if error.validator == "enum":
        return f"Invalid ADF{location}: value is not allowed."
    if error.validator == "type":
        return f"Invalid ADF{location}: expected {error.validator_value}."
    if error.validator == "minLength":
        return f"Invalid ADF{location}: expected a non-empty string."
    if error.validator == "minItems":
        return f"Invalid ADF{location}: expected at least one item."
    return f"Invalid ADF{location}: failed schema constraint '{error.validator}'."


def _most_specific_error(errors: list[ValidationError], value: Any) -> ValidationError:
    """Choose the schema error that best identifies the invalid ADF location."""

    candidates = _leaf_errors(errors)
    relevant = [error for error in candidates if error.validator not in {"anyOf", "oneOf"}]
    matching_type_error = next(
        (error for error in relevant if _error_matches_actual_type(error, value)),
        None,
    )
    if matching_type_error is not None:
        return matching_type_error
    return max(relevant or candidates, key=_error_specificity)


def _leaf_errors(errors: list[ValidationError]) -> list[ValidationError]:
    """Flatten nested anyOf/oneOf validation contexts to leaf errors."""

    leaves: list[ValidationError] = []
    for error in errors:
        if error.context:
            leaves.extend(_leaf_errors(list(error.context)))
        else:
            leaves.append(error)
    return leaves


def _json_pointer(path: Any) -> str:
    """Render a jsonschema path as an escaped JSON Pointer."""

    parts = [str(part).replace("~", "~0").replace("/", "~1") for part in path]
    return "/" + "/".join(parts) if parts else ""


def _error_specificity(error: ValidationError) -> int:
    """Score schema errors so deeper paths are considered more specific."""

    base = len(error.absolute_path)
    if error.validator == "required" and "/attrs" in _json_pointer(error.absolute_path):
        return base + 1
    return base


def _additional_property(message: str) -> str:
    """Extract an additional-property name from a jsonschema error message."""

    marker = "'"
    start = message.find(marker)
    end = message.find(marker, start + 1)
    return message[start + 1 : end] if start >= 0 and end > start else "unknown"


def _error_matches_actual_type(error: ValidationError, value: Any) -> bool:
    """Check whether a required-property error matches the node's actual type."""

    if error.validator != "required":
        return False

    parent_path = list(error.absolute_path)
    if parent_path[-1:] == ["attrs"]:
        parent_path = parent_path[:-1]
    parent = _value_at_path(value, parent_path)
    if (
        isinstance(parent, dict)
        and parent.get("type") == "link"
        and _required_property(error.message) == "href"
    ):
        return True

    schema_path = list(error.schema_path)
    if len(schema_path) < 2 or schema_path[0] != "definitions":
        return False

    definition = str(schema_path[1])
    expected_type = definition.removesuffix("_mark").removesuffix("_node")
    return isinstance(parent, dict) and parent.get("type") == expected_type


def _value_at_path(value: Any, path: list[Any]) -> Any:
    """Resolve a jsonschema path against a nested Python value."""

    current = value
    for part in path:
        if isinstance(current, list) and isinstance(part, int):
            current = current[part]
        elif isinstance(current, dict):
            current = current.get(part)
        else:
            return None
    return current


def _required_property(message: str) -> str:
    """Extract a required-property name from a jsonschema error message."""

    marker = "'"
    start = message.find(marker)
    end = message.find(marker, start + 1)
    return message[start + 1 : end] if start >= 0 and end > start else "unknown"
